test/functional/wallet_keypool_topup.py

```diff
@@ -24,71 +15,58 @@ def add_options(self, parser):
         self.add_wallet_options(parser)
 
     def set_test_params(self):
-        self.setup_clean_chain = True
-        self.num_nodes = 4
-        self.extra_args = [[], ['-keypool=100'], ['-keypool=100'], ['-keypool=100']]
+        self.num_nodes = 1
+        self.extra_args = [['-keypool=100']]
 
     def skip_test_if_missing_module(self):
         self.skip_if_no_wallet()
 
     def run_test(self):
-        wallet_path = self.nodes[1].wallets_path / self.default_wallet_name / self.wallet_data_filename
-        wallet_backup_path = self.nodes[1].datadir_path / "wallet.bak"
-        self.generate(self.nodes[0], COINBASE_MATURITY + 1)
-
-        self.log.info("Make backup of wallet")
-        self.stop_node(1)
-        shutil.copyfile(wallet_path, wallet_backup_path)
-        self.start_node(1, self.extra_args[1])
-        self.connect_nodes(0, 1)
-        self.connect_nodes(0, 2)
-        self.connect_nodes(0, 3)
-
-        for i, output_type in enumerate(["legacy", "p2sh-segwit", "bech32"]):
-
+        node = self.nodes[0]
+
+        output_types = ["legacy", "p2sh-segwit", "bech32"]
+        default_wallet = node.get_wallet_rpc(self.default_wallet_name)
+        for output_type in output_types:
+            wallet_name = str(output_type)
+            node.createwallet(wallet_name)
+            wallet = node.get_wallet_rpc(wallet_name)
             self.log.info("Generate keys for wallet with address type: {}".format(output_type))
-            idx = i+1
             for _ in range(90):
-                addr_oldpool = self.nodes[idx].getnewaddress(address_type=output_type)
+                addr_oldpool = wallet.getnewaddress(address_type=output_type)
             for _ in range(20):
-                addr_extpool = self.nodes[idx].getnewaddress(address_type=output_type)
+                addr_extpool = wallet.getnewaddress(address_type=output_type)
 
             # Make sure we're creating the outputs we expect
-            address_details = self.nodes[idx].validateaddress(addr_extpool)
-            if i == 0:
+            address_details = wallet.validateaddress(addr_extpool)
+            if output_type == "legacy":
                 assert not address_details["isscript"] and not address_details["iswitness"]
-            elif i == 1:
+            elif output_type == "p2sh-segwit":
                 assert address_details["isscript"] and not address_details["iswitness"]
             else:
                 assert not address_details["isscript"] and address_details["iswitness"]
 
-
             self.log.info("Send funds to wallet")
-            self.nodes[0].sendtoaddress(addr_oldpool, 10)
-            self.generate(self.nodes[0], 1)
-            self.nodes[0].sendtoaddress(addr_extpool, 5)
-            self.generate(self.nodes[0], 1)
+            default_wallet.send(outputs=[{addr_oldpool:10}, {addr_extpool:5}])
+            self.generate(node, 1)
```

Comment:
In bb54a7ad0cf22686e3f6735213bac1dd4453cb15 "test: use multiple wallets instead of nodes in `wallet_keypool_topup`"

The original method with separate transactions in separate blocks was the intended and correct behavior for this test. It needs to be that way so that the address with the lower index is found first to trigger the topup that will result in the second address being found. Since `send` shuffles the output order, we may get the second address before the first address which would result in an intermittent failure. Furthermore, even just separate transactions is not enough since the transactions could be in the incorrect order. Hence it needs to be separate transactions in separate blocks.

Reply:
That needs to be mentioned in the test case.